tests/unit/remote/test_http.py

```diff
@@ -13,3 +17,23 @@ def test_no_traverse_compatibility(dvc_repo):
 
     with pytest.raises(ConfigError):
         RemoteHTTP(dvc_repo, config)
+
+
+@pytest.mark.parametrize("response_code", [404, 403, 500])
+def test_download_fails_on_error_code(response_code, dvc_repo):
+    class ErrorStatusRequestHandler(BaseHTTPRequestHandler):
+        def do_GET(self):
+            self.send_response(response_code, message="Error message")
+            self.end_headers()
+
+    with StaticFileServer(ErrorStatusRequestHandler) as httpd:
+        url = "http://localhost:{}/".format(httpd.server_port)
+        config = {"url": url}
+
+        remote = RemoteHTTP(dvc_repo, config)
+        import os
+
+        with pytest.raises(HTTPErrorStatusCodeException):
+            remote._download(
+                URLInfo(os.path.join(url, "file.txt")), "file.txt"
```

Comment:
Simply use `URLInfo(url) / "file.txt"` or use `URLInfo` from the start. This will save you from an error of using `os.path` instead of `posixpath`.